### PR TITLE
implementing MVBA protocol

### PR DESCRIPTION
**Pull request by b00f**
This PR includes:
1- introducing Outgoing type. An outgoing type can be a direct message to a specific node, or a gossip message.
2. Updating error messages for MVBA
3. Implementation of MVBA based on Figure-3 of the spec (secure and Efficient Asynchronous Broadcast Protocols by Cachin)
4. Integrating the consensus protocols: (MVBA, ABBA and VCBC) 
5. `c-request` and `c-answer` messages for requesting a valid and signed proposal
6. Fixing the consensus failure in MVBA for the 3rd round case
7. Random  tests for the consensus protocol

### CODE REVIEW

**Review by davidrusu — APPROVED** [review]
Looks great. I left a few comments but I think none of them are blocking.

Fix the ones that are small enough for this PR and we can turn the rest into tickets

File: src/mvba/abba/mod.rs [hunk 1]
```diff
@@ -65,56 +70,61 @@ impl Abba {
...
-    pub fn pre_vote_one(&mut self, c_final: crate::mvba::vcbc::message::Message) -> Result<()> {
-        let justification = PreVoteJustification::FirstRoundOne(c_final);
+    pub fn pre_vote_one(&mut self, digest: Hash32, sig: Signature) -> Result<()> {
+        let justification = PreVoteJustification::WithValidity(digest, sig);
         self.pre_vote(Value::One, justification)
     }
 
     fn pre_vote(&mut self, value: Value, justification: PreVoteJustification) -> Result<()> {
+        if self.voted {
+            return Ok(());
```
[the remark this answers is not in the document]
Comment (b00f) [reply src/mvba/abba/mod.rs:80]: About this situation, 

It's normal that we receive a proposal after voting zero. so here is just prevent double voting. It's not an error in my point of view. I am not sure about buffering the votes. I add log here.

File: src/mvba/consensus.rs [hunk 2]
```diff
@@ -1,139 +1,309 @@
...
         }
-        // TODO:
-        // self.broadcaster.borrow_mut().take_bundles()
-        vec![]
     }
 
-    fn super_majority_num(&self) -> usize {
-        self.vcbc_map.len() - self.threshold
+    #[test]
+    fn test_random() {
```
[the remark this answers is not in the document]
Comment (b00f) [reply src/mvba/consensus.rs:243]: This test is really useful to fix some issues. Despite its name, the random here means picking a random message from the message buffer. I don't think quickcheck can be helpful here, even I thought about it before.

File: src/mvba/consensus.rs [hunk 3]
```diff
@@ -1,139 +1,309 @@
...
+            vcbc::MODULE_NAME => match bundle.target {
+                Some(target) => match self.vcbc_map.get_mut(&target) {
+                    Some(vcbc) => {
+                        let msg = bincode::deserialize(&bundle.payload)?;
+                        vcbc.receive_message(bundle.initiator, msg)?;
+                        if vcbc.is_delivered() {
+                            let abba = self.abba_map.get_mut(&target).unwrap();
+                            if abba.is_decided() && abba.decided_value() {
+                                self.decided_party = Some(target);
+                            }
```
[the remark this answers is not in the document]
Comment (b00f) [reply src/mvba/consensus.rs]: I put some comment, please check. Is it clear or need more explanation?

File: src/mvba/consensus.rs [hunk 4]
```diff
@@ -1,139 +1,309 @@
...
+                                    // abba is finished but still we don't have the proposal
+                                    // request it from the initiator
+                                    let data = vcbc::make_c_request_message(&self.id, target)?;
+
+                                    self.broadcaster.borrow_mut().send_to(
+                                        vcbc::MODULE_NAME,
+                                        Some(target),
+                                        data,
+                                        bundle.initiator,
+                                    );
```
[the remark this answers is not in the document]
Comment (b00f) [reply src/mvba/consensus.rs:144]: Yup, better we broadcast it instead to be in a safe side.

File: src/mvba/consensus.rs [hunk 5]
```diff
@@ -1,139 +1,309 @@
...
+                                }
+                            } else if !self.mvba.move_to_next_proposal()? {
+                                log::warn!("no more proposal");
+                            }
```
[the remark this answers is not in the document]
Comment (b00f) [reply src/mvba/consensus.rs:148]: Not sure if it is a good idea, in this case we might have double votes per proposal, unless we increase the sequence from 0 to 1.

File: src/mvba/consensus.rs [hunk 6]
```diff
@@ -1,139 +1,309 @@
...
-            if vcbc.is_delivered() {
-                delivered_count += 1;
+            if self.mvba.completed_vote() {
+                if let Some((proposal, sig)) = self.mvba.completed_vote_one() {
+                    let digest = Hash32::calculate(proposal);
+                    abba.pre_vote_one(digest, sig.clone())?;
+                }
+            } else {
+                abba.pre_vote_zero()?;
             }
```
[the remark this answers is not in the document]
Comment (b00f) [reply src/mvba/consensus.rs]: I put some comment, please check. Is it clear or need more explanation?

File: src/mvba/abba/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidrusu) [reply src/mvba/abba/mod.rs:80]: The idea was that the error is a signal to the caller that the proposal did not go through. They may want to bubble that message up to the client. i.e. "Consensus is in progress, retry your request later"

File: src/mvba/consensus.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidrusu) [reply src/mvba/consensus.rs]: Yes, makes sense 👍

File: src/mvba/consensus.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidrusu) [reply src/mvba/consensus.rs:154]: I see, you're worried about a tag re-use issue. Yeah good point. Ok, let's leave it as is for now and I'll create an issue to track this.

File: src/mvba/consensus.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidrusu) [reply src/mvba/consensus.rs]: Got it, yeah makes sense 👍

**Pull request merged**